### PR TITLE
Implement buffer adapter

### PR DESCRIPTION
**Pull request by tomskikh**

### CODE REVIEW

File: docs/source/savant_101/10_adapters.rst [hunk 1]
```diff
@@ -23,7 +23,7 @@ Adapters transfer video streams and metadata over network or locally. We impleme
 Savant Adapter Protocol
 -----------------------
 
-Savant uses a protocol based on `ZeroMQ <https://zeromq.org/>`__ and `Savant-RS <https://insight-platform.github.io/savant-rs/>`__ for communication between adapters and modules. It can be used to connect an adapter with other adapter, an adapter with a module, a module with a module, etc. The protocol is universal for source- and sink adapters.
+Savant uses a protocol based on `ZeroMQ <https://zeromq.org/>`__ and `Savant-RS <https://insight-platform.github.io/savant-rs/>`__ for communication between adapters and modules. It can be used to connect an adapter with other adapter, an adapter with a module, a module with a module, etc. The protocol is universal for source-, sink- and bridge adapters.
```
Comment (denisvmedyantsev) [inline docs/source/savant_101/10_adapters.rst]: "The protocol is universal for source-, sink- and bridge adapters." => "The protocol is universal for source, sink and bridge adapters."?

File: samples/buffer_adapter/demo.yml [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        # maximal lag in seconds
+        max_lag: 0.1
+        # lagged frames sequence length
+        lag_frames_interval: 500
+        # non-lagged frames sequence length
+        pass_frames_interval: 500
+
+    # primary detector element, inference is provided by the nvinfer Deepstream element
+    # model type is detector (other available types are: classifier, custom)
+    - element: nvinfer@detector
```
Comment (denisvmedyantsev) [inline samples/buffer_adapter/demo.yml]: Are you sure we need a model for the demo? I think the demo should be as clear and simple as possible. This is good for the user (understanding the main idea of the demo). And good for us (less support costs).

File: adapters/python/bridge/buffer.py [hunk 3]
```diff
@@ -0,0 +1,442 @@
...
+        elif buffer_size + QUEUE_ITEM_SIZE >= self._config.buffer.len:
+            self._buffer_is_full = True
+            self.logger.debug('Buffer is full, dropping the frame')
+            return False
+
+        self._queue.push(message_parts)
+        self.logger.debug('Pushed frame to the buffer')
+        return True
+
+    def push_service_message(self, message_parts: List[bytes]) -> bool:
```
Comment (denisvmedyantsev) [inline adapters/python/bridge/buffer.py:148]: Please explain what is the reason for the different handling of service and frame messages?

File: adapters/python/bridge/buffer.py [hunk 4]
```diff
@@ -0,0 +1,442 @@
...
+
+        self._socket_type, self._bind, self._socket = parse_zmq_socket_uri(
+            uri=config.zmq_sink_endpoint,
+            socket_type_enum=SenderSocketTypes,
+            socket_type_name=None,
+            bind=None,
+        )
+        assert (
+            self._socket_type == SenderSocketTypes.DEALER
+        ), 'Only DEALER socket type is supported for Egress'
```
Comment (denisvmedyantsev) [inline adapters/python/bridge/buffer.py:212]: It should be noted in the docs.

File: docs/source/savant_101/10_adapters.rst [hunk 5]
```diff
@@ -914,3 +914,70 @@ Running the adapter with the helper script:
...
+        -e BUFFER_LEN=1000 \
+        -e BUFFER_SERVICE_MESSAGES=100 \
+        -e BUFFER_THRESHOLD_PERCENTAGE=80 \
+        -e IDLE_POLLING_PERIOD=0.005 \
+        -e STATS_LOG_INTERVAL=60 \
+        -e METRICS_FRAME_PERIOD=1000 \
+        -e METRICS_TIME_PERIOD=10 \
+        -e METRICS_HISTORY=100 \
+        -e METRICS_PROVIDER=prometheus \
+        -e 'METRICS_PROVIDER_PARAMS={"port": 8000, "labels":{"adapter":"buffer"}}' \
```
Comment (denisvmedyantsev) [inline docs/source/savant_101/10_adapters.rst]: 'METRICS_PROVIDER_PARAMS={"port": 8000, "labels":{"adapter":"buffer"}}' 
=> 
METRICS_PROVIDER_PARAMS='{"port": 8000, "labels":{"adapter":"buffer"}}'?

File: adapters/python/bridge/buffer.py [hunk 6]
```diff
@@ -0,0 +1,442 @@
...
+                self._idle_polling_period,
+            )
+            time.sleep(self._idle_polling_period)
+            return None
+
+        message_parts = self._queue.pop(QUEUE_ITEM_SIZE)
+        if not message_parts[-1]:
+            message_parts.pop()
+        self._sent_messages += 1
+        frame_id = self._pipeline.add_frame('fps-meter', self._video_frame)
```
Comment (denisvmedyantsev) [inline adapters/python/bridge/buffer.py:254]: Add a frame and then delete it. Please explain.

File: samples/buffer_adapter/docker-compose.l4t.yml [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+      - METRICS_PROVIDER=prometheus
+      - METRICS_PROVIDER_PARAMS={"port":8000, "labels":{"adapter":"buffer"}}
+    command: python -m adapters.python.bridge.buffer
+
+  module:
+    image: ghcr.io/insight-platform/savant-deepstream-l4t:latest
+    restart: unless-stopped
+    volumes:
+      - zmq_sockets:/tmp/zmq-sockets
+      - ../../models/peoplenet_detector:/models
```
Comment (denisvmedyantsev) [inline samples/buffer_adapter/docker-compose.l4t.yml]: We changed paths for models/downloads (and new envs are required) in the last commit in the develop, please check and correct. (If a model remains in the module)

File: adapters/python/bridge/buffer.py (hunk 3 above)
Comment (tomskikh) [reply adapters/python/bridge/buffer.py:148]: When the buffer is full we drop messages. Service messages (EOS, Shutdown, etc.) we try not to drop cause they're important, eg. for cleaning resources after disconnecting the stream. For that we reserve some place in the buffer only for service.

File: adapters/python/bridge/buffer.py (hunk 6 above)
Comment (tomskikh) [reply adapters/python/bridge/buffer.py:254]: It's to measure FPS and log it to the console.

File: adapters/python/bridge/buffer.py (hunk 4 above)
Comment (tomskikh) [reply adapters/python/bridge/buffer.py:212]: It is noted in the docs.
https://github.com/insight-platform/Savant/pull/581/files#diff-46107dde0cc9a3ace43c076c303123111608509e69d8623fafc9a6e5f2ae10f5R939

File: samples/buffer_adapter/demo.yml (hunk 2 above)
Comment (tomskikh) [reply samples/buffer_adapter/demo.yml]: You're right, we don't need a model for this demo.

File: samples/buffer_adapter/docker-compose.l4t.yml (hunk 7 above)
Comment (tomskikh) [reply samples/buffer_adapter/docker-compose.l4t.yml]: Removed a model from the demo, now we don't need paths for model/downloads.

File: docs/source/savant_101/10_adapters.rst (hunk 1 above)
Comment (tomskikh) [reply docs/source/savant_101/10_adapters.rst]: FIxed.

File: docs/source/savant_101/10_adapters.rst (hunk 5 above)
Comment (tomskikh) [reply docs/source/savant_101/10_adapters.rst]: Fixed.

File: docs/source/savant_101/10_adapters.rst [hunk 8]
```diff
@@ -14,7 +14,7 @@ Currently, the single source- and multiple sink declarations are supported.
 Adapters
 ========
 
-The adapters are standalone programs executed in separate Docker containers. They communicate with modules via `ZeroMQ <https://zeromq.org/>`__: source adapters ingest data, and sink adapters consume data from modules.
+The adapters are standalone programs executed in separate Docker containers. They communicate with modules via `ZeroMQ <https://zeromq.org/>`__: source adapters ingest data, sink adapters consume data from modules and bridge adapters consume data from other sources and ingest it.
```
Comment (denisvmedyantsev) [inline docs/source/savant_101/10_adapters.rst:17]: I think the meaning of the sentence is unclear. The previous version was short, and "ingest" and "consume" refer to the module. Now this relation is broken. Could you please try to rephrase the sentence and make it clearer.

**Review by denisvmedyantsev — APPROVED** [review]

**Pull request merged**